src/Umbraco.Core/Composing/LightInject/LightInjectContainer.cs

```diff
@@ -56,9 +57,24 @@ protected static ServiceContainer CreateServiceContainer()
             // it is an index property. which means that eg protected or internal setters are OK.
             //Container.EnableAnnotatedPropertyInjection();
 
-            // ensure that we do *not* scan assemblies
-            // we explicitly RegisterFrom our own composition roots and don't want them scanned
-            container.AssemblyScanner = new AssemblyScanner(/*container.AssemblyScanner*/);
+            // ensure that we do *not* scan our assemblies
+            // but do not prevent users from doing so if they so desire.
+            bool PreventScanForUmbracoCompositionRootsPredicate(Type type, string s)
+            {
+                // Tests expect the other registered fallback to run, see RegisterAuto below.
+                if (!type.IsInterface)
+                    return false;
+
+                // The fallback factory method only applies to types in Umbraco namespaces
+                return type.FullName?.StartsWith("Umbraco", StringComparison.OrdinalIgnoreCase) ?? false;
+            }
+
+            // Explicitly registering a fallback factory method prevents ServiceContainer.GetEmitMethod
+            // from scanning the Umbraco assemblies for composition roots.
+            container.RegisterFallback(
+                PreventScanForUmbracoCompositionRootsPredicate,
+                request => throw new ResolveUnregisteredDependencyException(request.ServiceType)
```

Comment:
Do we want to throw an exception here? else can just be null and we don't have to worry about this accidentally throwing for weird reasons.

Reply:
I can argue it both ways, null from resolve is current behaviour so it's less of an impact which is nice, but i'm a big fan of fail hard/fast. When something goes wrong i'd much rather be thinking "oh there's something wrong with the ioc container/registrations let's go check that out" than "eh why have I got a random null reference exception".

I went with null at first but opted for an exception as I believe it is better, but happy to switch it to returning null.

Reply:
On second thoughts, LightInject.TryGetInstance is explicitly called which is documented as returning null on failure.

So I think returning null is better otherwise things will become very confusing.